server/src/main/java/org/elasticsearch/action/search/AbstractSearchAsyncAction.java

```diff
@@ -156,6 +156,16 @@ public final void executeNextPhase(SearchPhase currentPhase, SearchPhase nextPha
                     return;
                 }
             }
+            if (allowPartialResults == false && successfulOps.get() != getNumShards()) {
```

Comment:
We shouldn't have two tests to check if a partial failure occurred. This one could replace the block above and if successfulOps is less than the number of shards and shardSearchFailures.length is 0 we can assume that the failure was a shard unavailable exception ?

Reply:
Fixed in c2df1ab6996e2e35f5e7fc8bbf139fdadca09c93